pkg/syncer/k8scloudoperator/k8scloudoperator.go

```diff
@@ -147,6 +149,10 @@ func (k8sCloudOperator *k8sCloudOperator) GetPodVMUUIDAnnotation(ctx context.Con
 		return true, nil
 	})
 	if err != nil {
```

Comment:
When the polling times out, there will be no error since `nil` is being returned when annotation doesn't exist, and hence this if block will not be executed.

How about we don't change `if err != nil` block, and add another check for `if !exists` right after it. Something like:

```
if err != nil {
      return nil, logger.LogNewErrorf(log,
			"Unable to find pod %s and annotation %s on namespace %s in timeout: %d. Err: %+v",
			podName, vmUUIDLabel, podNamespace, timeout, err)
}
if !exists {
   // Return annotation NotFound error.
}

Reply:
makes sense. Done

Reply:
Actually I verified that polling time out returns error and confirmed the same with Balu. So, we changed it back to returning a bool indicating it is some other error which is not related to UUIDNOTFOUND. Hope this finalizes the error paths. With this change we can extend this to the additional error path we are adding in #1623 


Here is the testing details:
Polling time out returning error
```
Err: timed out waiting for the condition
```

RPC error with codes.NotFound:
```
Error: rpc error: code = NotFound desc = Unable to find vmware-system-vm-uuid annotation in pod with name: example-vanilla-block-pod on namespace: test-e2e.
```

Prometheus metrics will have:
```
vsphere_csi_volume_ops_histogram_sum{faulttype="csi.fault.nonstorage.VmUuidNotFound",namespace="unknown",optype="attach-volume",status="fail",voltype="block"} 541.137665148
vsphere_csi_volume_ops_histogram_count{faulttype="csi.fault.nonstorage.VmUuidNotFound",namespace="unknown",optype="attach-volume",status="fail",voltype="block"} 9
```



For other type of errors, we will continue to return Internal Fault:

CSI logs indicating it is other error like GetPod failed.
```
2022-03-11T08:39:18.393Z	ERROR	k8scloudoperator/k8scloudoperator.go:140	Failed to get the pod with name: example-vanilla-block-pod on namespace: test-e2e using K8s Cloud Operator informer. Err: pods "example-vanilla-block-pod" is forbidden: User "system:serviceaccount:vmware-system-csi:vsphere-csi-controller" cannot get resource "pods" in API group "" in the namespace "test-e2e"
sigs.k8s.io/vsphere-csi-driver/v2/pkg/syncer/k8scloudoperator.(*k8sCloudOperator).GetPodVMUUIDAnnotation.func1
	/build/pkg/syncer/k8scloudoperator/k8scloudoperator.go:140
...
...
```
CSI logs indicating it is other error like GetPod failed.
```
Error: rpc error: code = Unknown desc = Unable to find pod example-vanilla-block-pod and annotation vmware-system-vm-uuid on namespace test-e2e in timeout: 60000000000. Err: pods "example-vanilla-block-pod" is forbidden: User "system:serviceaccount:vmware-system-csi:vsphere-csi-controller" cannot get resource "pods" in API group "" in the namespace "test-e2e"
```

Prometheus metrics containing the InternalFault faulttype:
```
vsphere_csi_volume_ops_histogram_sum{faulttype="csi.fault.Internal",namespace="unknown",optype="attach-volume",status="fail",voltype="block"} 0.260274234
vsphere_csi_volume_ops_histogram_count{faulttype="csi.fault.Internal",namespace="unknown",optype="attach-volume",status="fail",voltype="block"} 9
```